azure-iot-device/azure/iot/device/provisioning/sk_provisioning_device_client.py

```diff
@@ -0,0 +1,87 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+"""
+This module contains one of the implementations of the Provisioning Device Client which uses Symmetric Key authentication.
+"""
+import logging
+from threading import Event
+import traceback
+from .provisioning_device_client import ProvisioningDeviceClient
+from .internal.polling_machine import PollingMachine
+
+logger = logging.getLogger(__name__)
+
+
+class SymmetricKeyProvisioningDeviceClient(ProvisioningDeviceClient):
+    """
+    Client which can be used to run the registration of a device with provisioning service
+    using Symmetric Key authentication.
+    """
+
+    def __init__(self, mqtt_state_based_provider):
+        """
+        Initializer for the Symmetric Key Registration Client
+        :param mqtt_state_based_provider: The state-based protocol provider. As of now this only supports MQTT.
+        """
+        super(SymmetricKeyProvisioningDeviceClient, self).__init__(mqtt_state_based_provider)
+        self._polling_machine = PollingMachine(mqtt_state_based_provider)
+        # To be defined by sample
+        self.on_registration_complete = None
+        self._polling_machine.on_disconnected = self._on_connection_state_change
+        self._polling_machine.on_registration_complete = self._on_device_registration_complete
+
+    def register(self):
+        """
+        Register the device with the provisioning service.
+        """
+        logger.info("Registering with Hub...")
+        register_complete = Event()
+
+        def callback_register(result=None, error=None):
+            # This could be a failed/successful registration result from the HUB
+            # or a error from polling machine. Response should be given appropriately
+            if result is not None:
+                if result.status == "assigned":
+                    logger.info("Successfully registered with Hub")
+                else:  # There be other statuses
+                    logger.error("Failed registering with Hub")
```

Comment:
>logger.error("Failed registering with Hub") [](start = 20, length = 43)

We should also probably throw an exception here, although admittedly, we have no coherent error story at all right now, so hard to say what error should be thrown.

This could be kicked down the road I suppose, but I don't think it'd hurt to throw a custom defined RegistrationError or something in the meantime.

Reply:
This is not really an exception. This is really the status check where something did not get registered. 
Exception may come handy in the next else block

Registration Error is a comletely different point however. It sounds like result but in the error mode

---
In reply to: [282618200](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r282618200) [](ancestors = 282618200)

Reply:
So on failure, we do nothing? I don't follow. Is this not so much a failure as it is the absence of success?

---
In reply to: [282632787](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r282632787) [](ancestors = 282632787,282618200)

Reply:
this is absence of success from the registration process...
there are statuses other than assigning assigned etc

---
In reply to: [282984448](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r282984448) [](ancestors = 282984448,282632787,282618200)